In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
from os import path
path="/content/drive/MyDrive/MachineLearningPraktikum/Praktikum7/data"

menginisialisasi lingkungan kerja untuk analisis data. Pertama, import pandas as pd mengimpor library pandas, yang sangat penting untuk manipulasi dan analisis data di Python, dan memberinya alias pd agar lebih mudah digunakan. Selanjutnya, kode me-mount (menghubungkan) Google Drive ke lingkungan runtime Google Colab menggunakan from google.colab import drive dan drive.mount('/content/drive'), yang memungkinkan kode mengakses file yang tersimpan di Google Drive pengguna. Terakhir, baris from os import path dan path="/content/drive/MyDrive/..." mengimpor modul untuk berinteraksi dengan sistem operasi (khususnya untuk path file) dan kemudian mendefinisikan sebuah variabel bernama path yang menunjuk ke lokasi spesifik dari direktori data dalam Google Drive pengguna, kemungkinan untuk digunakan dalam langkah pemuatan data berikutnya.

In [ ]:
df = pd.read_csv(path + "/dataset_satelit.csv")
df

No   Longitude  Lattitude     N     P      K    Ca    Mg      Fe  \
0      1  103.036658  -0.604417  2.64  0.15  0.415  0.51  0.31  119.96   
1      2  103.037201  -0.604689  2.75  0.17  0.568  0.76  0.58  102.63   
2      3  103.036359  -0.603012  1.77  0.12  0.339  0.49   0.6  107.37   
3      4  103.036950  -0.603219  2.30  0.15  0.460  0.74  0.67   96.02   
4      5  103.036802  -0.601969  2.05  0.14  0.308  0.64  0.72   87.01   
..   ...         ...        ...   ...   ...    ...   ...   ...     ...   
589  590  103.605867   0.057633  2.49  0.16  0.347  0.78  0.86   63.38   
590  591  103.606717   0.057100  2.74  0.15  0.466  0.73   0.5   51.04   
591  592  103.606250   0.056767  2.63  0.15  0.422  0.82  0.59   82.57   
592  593  103.606400   0.056517  2.75  0.17  0.502  0.69  0.53  102.07   
593  594  103.606033   0.056317  2.71  0.15  0.419  0.78  0.46   90.60   

         Mn  ...      b1  Sigma_VV  Sigma_VH      plia       lia      iafe  \
0    463.23  ...  0.0433   0.18183   0.04461  35.74446  35.79744  35.41161   
1    493.81  ...  0.0465   0.22079   0.04640  35.12096  35.14591  35.41510   
2    460.93  ...  0.0417   0.18926   0.03992  35.07724  35.07730  35.41135   
3    338.17  ...  0.0367   0.14769   0.03622  36.08078  36.08469  35.41583   
4    384.33  ...  0.0361   0.18205   0.03797  32.68855  32.69293  35.41592   
..      ...  ...     ...       ...       ...       ...       ...       ...   
589  269.95  ...  0.2336   0.13050   0.09390   0.12700   0.09860   0.02600   
590  683.42  ...  0.2506   0.21280   0.15920   0.20060   0.14730   0.03870   
591  396.18  ...  0.3413   0.27730   0.17820   0.25790   0.18690   0.04620   
592  246.35  ...  0.3413   0.32740   0.28760   0.29970   0.22250   0.05290   
593  371.46  ...  0.3482   0.41280   0.36380   0.38660   0.28990   0.06830   

     gamma0_vv  gamma0_vh  beta0_vv  beta0_vh  
0      0.22331    0.05479   0.31325   0.07686  
1      0.27116    0.05699   0.38033   0.07993  
2      0.23242    0.04902   0.32604   0.06876  
3      0.18138    0.04448   0.25440   0.06238  
4      0.22359    0.04664   0.31359   0.06541  
..         ...        ...       ...       ...  
589    0.00870    0.01690   0.00930   0.01630  
590    0.01380    0.02290   0.01270   0.01640  
591    0.01250    0.02250   0.01600   0.01930  
592    0.01800    0.03350   0.01790   0.01930  
593    0.02160    0.04460   0.02050   0.02220  

[594 rows x 34 columns]

#Import Data

inisialisasi penting dalam proyek Machine Learning (Pembelajaran Mesin), di mana ia mengimpor semua alat yang diperlukan untuk membangun dan mengevaluasi model. Ia memanggil Pandas (pd) untuk menangani dan memanipulasi data berbentuk tabel, dan NumPy (np) untuk melakukan operasi numerik yang efisien, terutama pada array. Dari pustaka Scikit-learn (sklearn)—standar untuk machine learning di Python—kode ini secara spesifik mengimpor tiga komponen utama: fungsi train_test_split untuk membagi data menjadi set pelatihan dan pengujian, kelas LinearRegression untuk membuat model regresi linier, dan fungsi r2_score serta mean_squared_error untuk mengukur dan menilai performa atau akurasi model yang sudah dibuat.

In [ ]:
import pandas as pd

import numpy as np

from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression

from sklearn.metrics import r2_score, mean_squared_error

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 594 entries, 0 to 593
Data columns (total 34 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   No         594 non-null    int64  
 1   Longitude  594 non-null    float64
 2   Lattitude  594 non-null    float64
 3   N          594 non-null    float64
 4   P          594 non-null    float64
 5   K          593 non-null    float64
 6   Ca         594 non-null    float64
 7   Mg         594 non-null    object 
 8   Fe         594 non-null    float64
 9   Mn         594 non-null    float64
 10  Cu         594 non-null    float64
 11  Zn         594 non-null    float64
 12  B          594 non-null    float64
 13  b12        594 non-null    float64
 14  b11        594 non-null    float64
 15  b9         594 non-null    float64
 16  b8a        594 non-null    float64
 17  b8         594 non-null    float64
 18  b7         594 non-null    float64
 19  b6         594 non-null    float64
 20  b5        

In [ ]:
df.describe()

No   Longitude   Lattitude           N           P           K  \
count  594.000000  594.000000  594.000000  594.000000  594.000000  593.000000   
mean   297.500000  106.878644   -1.024933    2.259091    0.141380    0.582175   
std    171.617307    4.949840    0.965349    0.395499    0.019782    0.222567   
min      1.000000  102.760857   -2.333750    1.140000    0.090000    0.122000   
25%    149.250000  102.927811   -2.233338    1.982500    0.130000    0.429000   
50%    297.500000  103.581969   -0.602276    2.280000    0.140000    0.549000   
75%    445.750000  113.403797   -0.257349    2.570000    0.150000    0.710000   
max    594.000000  113.434700    0.069251    3.230000    0.220000    1.489000   

               Ca          Fe           Mn          Cu  ...          b1  \
count  594.000000  594.000000   594.000000  594.000000  ...  594.000000   
mean     0.595094   74.613771   308.034697    2.391195  ...    0.177291   
std      0.366118   55.579655   241.731643    1.580296  ...    0.155615   
min      0.050000   21.080000     3.160000    0.090000  ...    0.014100   
25%      0.320000   40.705000   124.015000    1.172500  ...    0.046925   
50%      0.540000   65.650000   239.445000    2.225000  ...    0.072700   
75%      0.790000   87.372500   434.990000    3.357500  ...    0.318900   
max      2.820000  559.100000  2009.320000    8.170000  ...    0.751400   

         Sigma_VV    Sigma_VH        plia         lia        iafe   gamma0_vv  \
count  594.000000  594.000000  594.000000  594.000000  594.000000  594.000000   
mean     0.234474    0.102789   28.640422   28.664891   28.609569    0.202587   
std      0.070516    0.112310   15.325347   15.380384   15.329170    0.104357   
min      0.115170    0.021460    0.127000    0.098600    0.026000    0.008700   
25%      0.183210    0.039535   31.959745   31.968948   33.685353    0.183085   
50%      0.213385    0.046550   35.067930   35.110415   34.611565    0.233590   
75%      0.262242    0.059190   38.319135   38.441065   39.002760    0.271790   
max      0.512210    0.373000   47.592900   48.014640   39.209330    0.658960   

        gamma0_vh    beta0_vv    beta0_vh  
count  594.000000  594.000000  594.000000  
mean     0.051524    0.269642    0.062320  
std      0.012959    0.143728    0.024218  
min      0.016900    0.009300    0.016300  
25%      0.040250    0.244935    0.052772  
50%      0.050415    0.310380    0.068380  
75%      0.060410    0.364505    0.079020  
max      0.122300    0.814170    0.150620  

[8 rows x 33 columns]

Kode df.columns berfungsi untuk menampilkan daftar lengkap nama-nama kolom (variabel) yang ada di dalam DataFrame (tabel data) yang bernama df. Output yang dihasilkan—sebuah Index panjang berisi 'No', 'Longitude', 'Lattitude', variabel unsur hara ('N', 'P', 'K', dll.), dan banyak variabel lain seperti 'b1' hingga 'b12' serta variabel radar ('Sigma_VV', 'gamma0_vv', dll.)—secara keseluruhan memberikan gambaran tentang fitur-fitur yang tersedia dalam dataset tersebut. Ini adalah langkah diagnostik dasar untuk memahami struktur data sebelum melakukan pemrosesan, analisis, atau pemodelan lebih lanjut.

In [ ]:
df.columns

Index(['No', 'Longitude', 'Lattitude', 'N', 'P', 'K', 'Ca', 'Mg', 'Fe', 'Mn',
       'Cu', 'Zn', 'B', 'b12', 'b11', 'b9', 'b8a', 'b8', 'b7', 'b6', 'b5',
       'b4', 'b3', 'b2', 'b1', 'Sigma_VV', 'Sigma_VH', 'plia', 'lia', 'iafe',
       'gamma0_vv', 'gamma0_vh', 'beta0_vv', 'beta0_vh'],
      dtype='object')

In [ ]:
#ubah kolom Mg ke float
df['Mg'] = pd.to_numeric(df['Mg'], errors='coerce')

#hilangkan baris yang ada missing value
df = df.dropna()

In [ ]:
# x = df[['b2', 'b3', 'b2', 'b1', 'Sigma_VV', 'Sigma_VH']]

# y = df['N']

In [ ]:
x = df[['P', 'K', 'Ca', 'Mg', 'Fe', 'Mn', 'Cu', 'Zn', 'B', 'b12', 'b11', 'b9', 'b8a', 'b8', 'b7', 'b6', 'b5', 'b4', 'b3', 'b2', 'b1', 'Sigma_VV', 'Sigma_VH', 'plia', 'lia', 'iafe', 'gamma0_vv', 'gamma0_vh', 'beta0_vv', 'beta0_vh']]

y= df['N']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    x, y, test_size=0.2, random_state=42
)

In [ ]:
model = LinearRegression()

model.fit(X_train, y_train)

LinearRegression()

1. Pembersihan Data: Baris pertama mengubah kolom 'Mg' menjadi tipe data numerik (pd.to_numeric), dengan opsi errors='coerce' yang akan mengubah nilai yang tidak valid (non-numerik) menjadi Missing Value (NaN). Setelah itu, baris df = df.dropna() menghapus semua baris dalam DataFrame (df) yang mengandung nilai Missing Value (NaN) di kolom mana pun, memastikan data bersih dari entri yang rusak atau hilang.

2. Pemilihan Fitur (X) dan Target (Y): Kode ini kemudian mendefinisikan variabel independen (X) yang akan digunakan untuk memprediksi, dan variabel dependen (y) yang merupakan target prediksi. Terdapat dua blok pemilihan X dan y, namun pada akhirnya blok kedua yang panjang mendefinisikan X sebagai kumpulan semua fitur (unsur hara, band citra, dan variabel radar) dan y sebagai kolom 'N' (kemungkinan unsur hara Nitrogen) yang akan diprediksi.

3. Pembagian Data: Fungsi train_test_split membagi pasangan data X dan y menjadi empat bagian: X_train dan y_train untuk melatih model, serta X_test dan y_test untuk menguji performa model, dengan test_size=0.1 berarti 10% data dialokasikan untuk pengujian.

4. elatihan Model: Akhirnya, sebuah objek LinearRegression dibuat dan disimpan dalam variabel model, lalu model tersebut dilatih (dipelajari) menggunakan data pelatihan (X_train dan y_train) melalui perintah model.fit().

In [ ]:
y_pred = model.predict(X_test)

#evaluasi model

r2 = r2_score(y_test, y_pred)

mse = mean_squared_error(y_test, y_pred)

rmse = np.sqrt(mse)

print("R² Score:", r2)

print("RMSE:", rmse)

R² Score: 0.7549955124484078
RMSE: 0.19392193397227142


kode ini menggunakan model regresi linier yang sudah dilatih di langkah sebelumnya untuk membuat prediksi, lalu menghitung seberapa baik performa model tersebut. Pertama, y_pred = model.predict(X_test) meminta model memprediksi nilai target (y) berdasarkan data pengujian yang belum pernah dilihat (X_test). Kemudian, bagian evaluasi menghitung dua metrik penting:r2 = r2_score(y_test, y_pred): Menghitung $R^2$ Score (Koefisien Determinasi), yang mengukur proporsi variabilitas dalam variabel dependen (target) yang dapat dijelaskan oleh model. Nilai $R^2$ yang dicetak sebesar 0.755 menunjukkan model dapat menjelaskan sekitar 75.5% dari variasi data.mse = mean_squared_error dan rmse = np.sqrt(mse): Menghitung $MSE$ (Mean Squared Error), yang mengukur rata-rata kuadrat kesalahan antara nilai sebenarnya (y_test) dan nilai prediksi (y_pred), lalu menghitung $RMSE$ (Root Mean Squared Error) dengan mengakarkannya. Nilai $RMSE$ sebesar 0.194 menunjukkan rata-rata besarnya kesalahan prediksi model dalam unit yang sama dengan variabel target ('N').

In [ ]:
coeff = pd.DataFrame({
    "Fitur": X_train.columns,
    "Koefisien": model.coef_
})

print(coeff)

membuat sebuah tabel baru (DataFrame) bernama coeff yang digunakan untuk memvisualisasikan Koefisien Regresi dari model yang sudah dilatih. Tabel ini memiliki dua kolom: kolom "Fitur" yang berisi semua nama variabel independen (diambil dari X_train.columns), dan kolom "Koefisien" yang berisi nilai koefisien regresi yang dihitung oleh model (model.coef_). Nilai koefisien ini menunjukkan seberapa besar perubahan pada variabel target ('N') untuk setiap kenaikan satu unit pada fitur tertentu, dengan asumsi fitur lain konstan. Dengan mencetak tabel coeff, kita bisa melihat fitur mana yang memiliki pengaruh positif terbesar, pengaruh negatif terbesar, atau pengaruh yang sangat kecil terhadap prediksi model.

In [ ]:
import statsmodels.api as sm

X_sm = sm.add_constant(x)

# tambahkan intercept

model_ols = sm.OLS(y, X_sm).fit()

print(model_ols.summary())

Mengimpor Alat Statistik: Baris import statsmodels.api as sm mengimpor pustaka statsmodels, yang merupakan pilihan populer di kalangan ahli statistik karena output yang sangat detail.Menambahkan Intercept (Konstanta): Baris X_sm = sm.add_constant(X) secara eksplisit menambahkan kolom konstanta (yang berisi angka 1) ke variabel independen X. Dalam analisis regresi, kolom ini sangat penting karena merepresentasikan nilai intercept  pada persamaan regresi (nilai prediksi Y ketika semua nilai X adalah nol).Melatih Model OLS: Baris model_ols = sm.OLS(y, X_sm).fit() membuat dan melatih model OLS (Ordinary Least Squares), yang merupakan metode untuk Regresi Linier, menggunakan data target (y) dan data fitur yang sudah ditambahkan konstanta (X_sm).Menampilkan Ringkasan Statistik: Terakhir, print(model_ols.summary()) mencetak ringkasan statistik yang komprehensif dari hasil regresi. Ringkasan ini biasanya mencakup koefisien,  Score, nilai p-value untuk menguji signifikansi setiap fitur, dan berbagai metrik diagnostik lainnya.